sdk/iot/Azure.Iot.Hub.Service/tests/IotHubSasCredentialTests.cs

```diff
@@ -0,0 +1,107 @@
+
+        static IotHubSasCredentialTests()
+        {
+            // Create a base64 encoded key to be used for mocking the shared access key value.
+            var rnd = new Random();
+            var rndBytes = new byte[32];
+            rnd.NextBytes(rndBytes);
+
+            s_testSharedAccessKey = Convert.ToBase64String(rndBytes);
+        }
+
+        [Test]
+        public void ValidateCtorParameters()
+        {
+            // Arrange
+            string sharedAccessPolicy = "policy";
+            string sharedAccessKey = "key";
+
+            // Act and Assert
+            Assert.Throws<ArgumentNullException>(() => new IotHubSasCredential(sharedAccessPolicy, null));
+            Assert.Throws<ArgumentNullException>(() => new IotHubSasCredential(null, sharedAccessKey));
+        }
+
+        [Test]
+        public void CtorParametersNegativeTimeToLiveThrowsException()
+        {
+            // Arrange
+            string sharedAccessPolicy = "policy";
+            string sharedAccessKey = "key";
+            var ttl = TimeSpan.FromSeconds(-10);
+
+            // Act and Assert
+            Assert.Throws<ArgumentException>(() => new IotHubSasCredential(sharedAccessPolicy, sharedAccessKey, ttl));
+        }
+
+        [Test]
+        public void CtorSetsDefaultTimeToLive()
+        {
+            // Arrange
+            string sharedAccessPolicy = "policy";
+            string sharedAccessKey = "key";
+
+            // Act
+            var credentials = new IotHubSasCredential(sharedAccessPolicy, sharedAccessKey);
+
+            // Assert
+            credentials.SasTokenTimeToLive.Should().BePositive();
+        }
+
+        [Test]
+        public async Task CachesTokenIfNotExpired()
+        {
+            // Arrange
+            var ttl = TimeSpan.FromMinutes(1);
+
+            // Act
+            var credential = new IotHubSasCredential("policy", s_testSharedAccessKey, ttl)
+            {
+                Endpoint = s_endpoint,
+            };
+
+            string initialToken = credential.GetSasToken();
+            await Task.Delay(2000);
+            string newToken = credential.GetSasToken();
+
+            // Assert
+            newToken.Should().BeEquivalentTo(initialToken, "Token should be cached and returned");
+        }
+
+        [Test]
+        public async Task RegeneratesTokenIfExpired()
+        {
+            // Arrange
+            var ttl = TimeSpan.FromSeconds(1);
+
+            // Act
+            var credential = new IotHubSasCredential("policy", s_testSharedAccessKey, ttl)
+            {
+                Endpoint = s_endpoint,
+            };
+
+            string initialToken = credential.GetSasToken();
+            await Task.Delay(2000);
+            string newToken = credential.GetSasToken();
+
+            // Assert
+            newToken.Should().NotBeEquivalentTo(initialToken, "Token has expired, and should be regenerated");
+        }
+    }
```

Comment:
Is it worthwhile adding tests that verify that token is refreshed when at 15% or less of it's lifespan?